File: solutions/csharp/log-levels/1/LogLevels.cs
using System.Text.RegularExpressions;

static class LogLine
{
    public static string Message(string logLine) => Parse(logLine).Message;

    public static string LogLevel(string logLine) => Parse(logLine).Level.ToLower();

    public static string Reformat(string logLine) =>
        $"{Message(logLine)} ({LogLevel(logLine)})";

    private static (string Level, string Message) Parse(string logLine)
    {
        var match = Regex.Match(logLine, @"^\[(?<level>INFO|WARNING|ERROR)\]:\s*(?<message>.*)$");

        return (
            match.Groups["level"].Value,
            match.Groups["message"].Value.Trim());
    }
}

File: solutions/fsharp/log-levels/1/LogLevels.fs
module LogLevels

let private separatorIndex (logLine: string) = logLine.IndexOf "]:"

let message (logLine: string): string =
    let separator = separatorIndex logLine
    logLine.Substring(separator + 2).Trim()

let logLevel (logLine: string): string =
    let separator = separatorIndex logLine
    logLine.Substring(1, separator - 1).ToLowerInvariant()

let reformat (logLine: string): string =
    sprintf "%s (%s)" (message logLine) (logLevel logLine)
